sys/net/gnrc/transport_layer/tcp/gnrc_tcp_eventloop.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (C) 2015 Simon Brummer
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_gnrc
+ * @{
+ *
+ * @file
+ * @brief       GNRC's TCP event processing loop
+ *
+ * @author      Simon Brummer <brummer.simon@googlemail.com>
+ * @}
+ */
+
+#include <utlist.h>
+#include <errno.h>
+#include "net/gnrc/pkt.h"
+#include "net/conn/tcp.h"
+#include "net/gnrc/tcp.h"
+#include "net/gnrc/tcp/hdr.h"
+#include "tcp_internal/pkt.h"
+#include "tcp_internal/fsm.h"
+#include "tcp_internal/helper.h"
+#include "tcp_internal/option.h"
+#include "tcp_internal/eventloop.h"
+
+#ifdef MODULE_GNRC_IPV6
+#include "net/gnrc/ipv6.h"
+#endif
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/**
+ * @brief send function, used to pass paket down the network stack
+ *
+ * @param[in] pkt   paket to pass down the network stack
+ *
+ * @return   zero on success
+ * @return   negative value on error
+ * @return   -EBADMSG if tcp header is missing
+ */
+static int8_t _send(gnrc_pktsnip_t *pkt)
+{
+    /* NOTE: Sending Direction: pkt = nw, nw->next = tcp, tcp->next = payload */
+    gnrc_pktsnip_t *tcp = NULL;
+
+    /* Search for tcp header */
+    LL_SEARCH_SCALAR(pkt, tcp, type, GNRC_NETTYPE_TCP);
+    if (tcp == NULL) {
+        DEBUG("gnrc_tcp_eventloop : _send() : tcp header missing.\n");
+        gnrc_pktbuf_release(pkt);
+        return -EBADMSG;
+    }
+
+    /* Dispatch to network layer */
+    if (!gnrc_netapi_dispatch_send(pkt->type, GNRC_NETREG_DEMUX_CTX_ALL, pkt)) {
+        DEBUG("gnrc_tcp_eventloop : _send() : network layer not found\n");
+        gnrc_pktbuf_release(pkt);
+    }
+    return 0;
+}
+
+/**
+ * @brief recv function, used to call fsm on packet reception
+ *
+ * @param[in] pkt   incomming paket to process
+ *
+ * @return zero on success
+ * @return negative value on error
+ * @return -EACCES if not able to get write access to packet
+ * @return -ERANGE if segment offset is less than 5
+ * @return -ENOMSG if paket can't be marked
+ * @return -EINVAL if checksum was invalid
+ * @return -ENOTCONN if no module is interested in this context
+ */
+static int8_t _receive(gnrc_pktsnip_t *pkt)
+{
+    /* NOTE: Receiving direction: pkt = payload, payload->next = tcp, tcp->next = nw */
+    uint16_t ctl = 0;
+    uint16_t src = 0;
+    uint16_t dst = 0;
+    uint8_t hdr_size = 0;
+    uint8_t syn = 0;
+
+    gnrc_pktsnip_t *tcp = NULL;
+    gnrc_pktsnip_t *ip6 = NULL;
+    gnrc_pktsnip_t *rst = NULL;
```

Comment:
Reset: In case the there is nobody listening on that Port or the Connection is already open, there must be a direct reply with a Reset packet